feat: expand security documentation (#110)
canonical/charmed-aether-sd-core: docs/explanation/security.md

Reviewer: Our security claims are too high and we are too verbose.

Each Charmed Aether SD-Core charm generates its private key and a certificate signing request (CSR). The CSR is then transmitted to a TLS certificate provider, which in turn signs the certificate and sends it back to the charm. Subsequently, the 5G network function utilizes this certificate to encrypt its communications with other network functions.

By default, the TLS certificate provider employed is the [self-signed-certificates operator](https://charmhub.io/self-signed-certificates). However, users have the flexibility to utilize any TLS certificates provider as long as it supports the `tls-certificates` integration.

## Access Control and Secure Communication in NMS

> The Charmed Aether SD-Core uses JWT-based authentication with role-based permissions to ensure secure access. Admin users have full control over all resources and accounts, while all other users have restricted permissions, limited to managing the resources they create and their own accounts. Authentication tokens expire after one hour, adding an extra layer of security. Furthermore, all HTTP endpoints in the Network Management System (NMS) are secured with HTTPS to ensure encrypted communication and prevent unauthorized interception.

Author: Done

> The Charmed Aether SD-Core uses JWT-based authentication with role-based permissions to ensure secure access. Admin users have full control over all resources and accounts, while all other users have restricted permissions, limited to managing the resources they create and their own accounts. Authentication tokens expire after one hour, adding an extra layer of security. Furthermore, all HTTP endpoints in the Network Management System (NMS) are secured with HTTPS to ensure encrypted communication and prevent unauthorized interception.

Author: Done

## Database Security

> The Charmed Aether SD-Core ensures security and isolation by keeping network functions, NMS users and subscribers' data in different databases.

Author: Done

Granular access control restricts each database user's permissions to their respective data store, ensuring isolation, minimizing risks and enforcing the least privilege principle.

## Input Validation

Input validation mechanisms are implemented within charms and the Network Management System (NMS) to enhance the secure handling of user-provided inputs. These validations primarily target inputs used for:

- Configuring the Network Management System (NMS)
- Managing and operating charms

By ensuring validation and sanitization of certain inputs, the platform aims to reduce the risks associated with vulnerabilities like SQL injection, command injection, and exploitation attempts.

## Log Confidentiality

> The Charmed Aether SD-Core adopts enhanced log confidentiality practices to ensure that sensitive or confidential data is excluded from log files.

Reviewer: ```suggestion
Charmed Aether SD-Core adopts enhanced log confidentiality practices to ensure that sensitive or confidential data is excluded from log files.
```

Author: Done

- Subscriber authentication tokens or identifiers such as location
- Encryption keys
- Configuration secrets or network-specific credentials

By adhering to these practices, the platform prevents unintended disclosure of confidential information through logging mechanisms and ensures privacy.

## Secure Dependencies

> All libraries and dependencies utilized in Juju charms are continuously monitored, scanned, and updated using **Renovate**. This automation ensures:

Reviewer: ```suggestion
All libraries and dependencies utilized in Juju charms are continuously monitored, scanned, and updated. This automation ensures:
```

Author: Done

- The use of up-to-date libraries and dependencies.
- Identification and mitigation of vulnerabilities in third-party packages.

> By leveraging Renovate, the platform ensures that all Juju charms remain secure and aligned with the latest updates, reducing risks associated with outdated or vulnerable code.

Reviewer: ```suggestion
```

Author: Removed.

Each Charmed Aether SD-Core charm generates its private key and a certificate signing request (CSR). The CSR is then transmitted to a TLS certificate provider, which in turn signs the certificate and sends it back to the charm. Subsequently, the 5G network function utilizes this certificate to encrypt its communications with other network functions.

By default, the TLS certificate provider employed is the [self-signed-certificates operator](https://charmhub.io/self-signed-certificates). However, users have the flexibility to utilize any TLS certificates provider as long as it supports the `tls-certificates` integration.

> ## Access Control and Secure Communication in NMS

Reviewer: This whole block is very verbose and can likely be reduced to maybe 3 sentences on authentication, authorization, and access

> Charmed Aether SD-Core platform implements strict access controls and role-based permissions. Other than the admin user, all users are granted limited permissions, ensuring that they can only access, edit, or delete the resources they have created. These include slices, device groups and subscribers. Additionally, each user can only view and manage their own account and has ability to change their own password.

Reviewer: `strict access controls` -> I recommend we are more precise, for example: JWT based authentication.

Author: OK.

> Charmed Aether SD-Core platform implements strict access controls and role-based permissions. Other than the admin user, all users are granted limited permissions, ensuring that they can only access, edit, or delete the resources they have created. These include slices, device groups and subscribers. Additionally, each user can only view and manage their own account and has ability to change their own password.

Reviewer: `can only access, edit, or delete the resources they have created`? Are you sure that's correct?

Author: The above sentence briefly says that they can create resource, then access (this can be replaced by view), edit and delete. I do not see a problem here. Do you think am i missing something ?

> Charmed Aether SD-Core platform implements strict access controls and role-based permissions. Other than the admin user, all users are granted limited permissions, ensuring that they can only access, edit, or delete the resources they have created. These include slices, device groups and subscribers. Additionally, each user can only view and manage their own account and has ability to change their own password.

Author: Here, NMS supports 2 roles: admin and non-admin.  The first user is created with admin role which has ability to do almost everthing except deleting its own account. The non-admin users are granted with limited permissions as described in the paragraph above.

The admin user, however, has full control over the platform. These capabilities include creating and deleting users, managing all resources (slices, device groups, subscribers, etc.), and viewing and editing all accounts. This hierarchical access control provides a secure and organized way to manage users and resources within the system.

To further enhance security, authentication tokens are valid for only one hour. This ensures that access to the platform remains temporary and reduces the risk of unauthorized access in case a token is exposed. This layered security model ensures that permissions are strictly enforced, reducing potential risks and maintaining the integrity of the Charmed Aether SD-Core platform.

> Additionally, all HTTP endpoints in the Network Management System (NMS) are secured using HTTPS. This encrypts data in transit and mitigates risks of man-in-the-middle (MITM) attacks. With strong TLS implementations, HTTPS ensures secure connections and protects sensitive information transmitted within the system.

Reviewer: This is too verbose in my opinion, no need to explain what HTTPS is.

Author: OK, I can make it more concise.

## Database Security

> To ensure enhanced security and organization, the Charmed Aether SD-Core platform implements three distinct and isolated data stores. Each data store is dedicated to a specific type of data:

Reviewer: We don't call it "The Charmed Aether SD-Core platform" - we would use "Charmed Aether SD-Core" instead.

Author: OK, Done.

> To ensure enhanced security and organization, the Charmed Aether SD-Core platform implements three distinct and isolated data stores. Each data store is dedicated to a specific type of data:

Reviewer: A similar point here, I think we are unnecessarily verbose

- Network Functions (NFs) data: Stores information related to configurations, operations and the policies of 5G Core network.
- Subscriber data: Contains authentication data and other details related to subscribers.
- NMS users data: Manages data specific to the Network Management System (NMS), including user information, roles, and access permissions.

Access to these databases is tightly controlled by granting different database users the necessary permissions to only access their respective data stores. This granular access control ensures that each database is securely isolated, which reduces risks, prevents unauthorized access, and enforces the principle of least privilege.

## Input Validation

> Comprehensive input validation is implemented within charms and the Network Management System (NMS) to mitigate the risk of injection attacks and ensure secure handling of user-provided inputs. This validation applies to inputs used for:

Reviewer: `Comprehensive input validation`: This is an overstatement looking at the actual input validation done in the upstream project.

Author: The section is updated.

- Configuring the Network Management System (NMS)
- Managing and operating charms

By performing thorough validation and sanitization of inputs in these components, the platform effectively prevents vulnerabilities such as SQL injection, command injection, and other malicious exploitation attempts. This ensures the integrity and security of the platform's underlying systems and protects against unauthorized access or data manipulation.

## Log Confidentiality

> The Charmed Aether SD-Core platform enforces strict log confidentiality standards to ensure that sensitive or confidential data is never included in log files. This policy applies to all Network Function logging frameworks and ensures the exclusion of:

Reviewer: `strict log confidentiality standards`: Again, this is an overstatement, the upstream community doesn't enforce such strict standards,

Author: The section is updated.

- Network Functions (NFs) data: Stores information related to configurations, operations and the policies of 5G Core network.
- Subscriber data: Contains authentication data and other details related to subscribers.
- NMS users data: Manages data specific to the Network Management System (NMS), including user information, roles, and access permissions.

Access to these databases is tightly controlled by granting different database users the necessary permissions to only access their respective data stores. This granular access control ensures that each database is securely isolated, which reduces risks, prevents unauthorized access, and enforces the principle of least privilege.

## Input Validation

Comprehensive input validation is implemented within charms and the Network Management System (NMS) to mitigate the risk of injection attacks and ensure secure handling of user-provided inputs. 

This validation applies to inputs used for:

- Configuring the Network Management System (NMS)
- Managing and operating charms

> By performing through validation and sanitization of inputs in these components, the platform effectively prevents vulnerabilities such as SQL injection, command injection, and other malicious exploitation attempts. This ensures the integrity and security of the platform's underlying systems and protects against unauthorized access or data manipulation.

Reviewer 2: ```suggestion
By performing thorough validation and sanitation of inputs in these components, the platform effectively prevents vulnerabilities such as SQL injection, command injection, and other malicious exploitation attempts. This ensures the integrity and security of the platform's underlying systems and protects against unauthorized access or data manipulation.
```

Author: I do not think that replacing sanitization with sanitation is reasonable here.
Sanitization has a meaning that cleaning, filtering or preparing something to make it safe which is generally used in techincal programming context.  However, the  sanitation has a meaning of hygiene and cleaning something in the pyhsical world. Hence, I would like to keep existing word.
